In [1]:
import numpy as np
from matplotlib import pyplot as plt
from scipy import linalg
import scipy

In [2]:
plt.style.use('seaborn-poster')

### Pregunta 1

Dividimos el problema en cuatro partes, una para cada bloque. Asumimos que $x$ e $y$ están en la direcciones de la cuñeta, con $x$ siendo la dirección horizontal e $y$ siendo la dirección vertical.

#### Masa 1

Primero resolvemos el problema de la masa $m_1$: Claramente tenemos la fuerza de la tensión la cual es $T_1$, tenemos el peso $m_1g$, la fuerza normal en sentido perpendicular a la superficie $N_{1}$, y por último la fricción con fuerza $\mu_{1}N_1$.

Ahora ordenamos las ecuaciones, primero creamos la ecuación del eje $x$:
$$T_1+Fr-m_1cos(\theta)=T_1+N_{1}\mu_{1}-m_1cos(\theta)g=am_1$$

Y la ecuación del eje $y$:
            $$N_{1}-m_{1}gsen(\theta)=0$$
            
Resolvemos y sacamos $N_1$:
            $$N_{1}=m_{1}gsen(\theta)$$

Reemplazamos este $N_1$ en la ecuación de $x$
            $$T_1+m_{1}g\mu_{1}sen(\theta)-m_{1}gcos(\theta)=am_1$$

#### Masa 2

Para la masa $m_2$ tenemos las mismas fuerzas que en $m_1$ salvo que hay otra fuerza de tensión adicional. Entonces, tenemos la tension de arriba: $T_2$, tenemos la tensión de abajo $-T_1$. La otra fuerza horizontal que hay es el roce con $\mu_{2}$, la cual es $-N_2\mu_{2}$. En resumen para el eje $x$:
        $$T_2-T_1+N_2\mu_{2}-m_2gcos(\theta)=m_2a$$

Ahora para el eje $y$, tenemos la fuerza del peso con $-m_{2}g$ y la fuerza de la normal con $N_2$ resumiendo obtenemos que para el eje $y$ la siguiente ecuación:
            $$N_{2}=m_{2}gsen(\theta)$$

Haciendo analogo a la masa $m_1$ reemplazamos $N_2$ en la ecuacion de $x$:
                $$T_{2}-T_1+m_{2}g\mu_{2}sen(\theta)-m_2gcos(\theta)=m_2a$$

#### Masa 3

Para la masa $m_3$ tenemos lo mismo que para $m_2$ pero para $m_3$.

La ecuación horizontal tiene a la fuerza de la Tensión pero para $T_3$, y la fuerza de la tensión de $T_2$. Tambien tenemos el roce con $\mu_{3}$. Entonces tenemos que:
        $$T_3-T_2+N_3\mu_{3}-m_3gcos(\theta)=m_3a$$

Para la ecuación de $y$ es lo mismo que para la masa $m_2$ pero con $N_3$ y $m_3$.
                $$N_3=m_3gsen(\theta)$$

Reemplazando $N_3$ en la ecuación de $x$:
            $$T_3-T_2+m_3g\mu_{3}sen(\theta)-m_3gcos(\theta)=m_3a$$

#### Masa 4

En este caso para $m_4$ no hay ecuación de $x$ porque no hay movimiento horizontal, ni hay fuerzas horizontales, así que solo hay ecuación en el eje $y$.
        $$T_3-m_{4}g=-am_4$$

#### Planteamiento del problema

Vamos a ver el problema de la siguiente forma, dado que tenemos $\vec{m}$, y nos falta por encontrar $\vec{T}$, podemos escribirlo de forma que las masas estén en un lado y las tensiones en otro, entonces tomamos las masas como el vector $\vec{b}$ y el vector de Tensión $\vec{T}$ como el equivalente al vector $\vec{x}$. Entonces escribimos el siguiente sistema de ecuaciones.
$$A\vec{T}=\vec{m}$$

Definimos $\vec{T}$ de la siguiente forma:
$$\vec{T}=\begin{bmatrix}
        T_1 \\
        T_2 \\
        T_3\\
        a
        \end{bmatrix}$$

Reordenamos las ecuaciones para que que se multipliquen con $\vec{T}$, así tratamos de dejar las tensiones y la aceleración de un lado y las masas en otra.

Tenemos que la primera fila de $A\vec{T}$ es:
$$T_1-am_{1}=-m_1g\mu_{1}sen(\theta)+m_{1}gcos(\theta)$$

La segunda fila de $A\vec{T}$ es:
$$T_{2}-T_1-m_2a=-m_{2}g\mu_{2}sen(\theta)+m_2gcos(\theta)$$

La tercera fila de $A\vec{T}$ es:
$$T_3-T_2-m_3a=-m_3g\mu_{3}sen(\theta)+m_3gcos(\theta)$$

La ultima fila de $A\vec{T}$ es:
$$T_3+am_4=m_{4}g$$

Entonces $A\vec{T}$ queda de la siguiente forma:
$$A\vec{T}=\begin{bmatrix}
        T_1 & 0 & 0 &-am_{1}\\
        -T_{1}& T_2& 0 &-m_2a \\
        0 & -T_2 & T_3 &-m_3a \\
        0 & 0 & T_3 & m_4a
        \end{bmatrix}$$

Escribimos A entonces como:
$$A=\begin{bmatrix}
        1& 0 & 0 & -m_1 \\
        -1& 1& 0 &-m_2\\
        0 & -1 & 1 &-m_3 \\
        0 & 0 & 1 & m_4
        \end{bmatrix}$$

Teniendo el vector $\vec{b}$:
$$\vec{b}=\begin{bmatrix}
       -m_1g\mu_{1}sen(\theta)+m_{1}gcos(\theta) \\
       -m_2g\mu_{2}sen(\theta)+m_2gcos(\theta)\\
       -m_3g\mu_{3}sen(\theta)+m_3gcos(\theta)\\
       m_4g
       \end{bmatrix}
$$

Dado que $\theta=45°$, entonces $cos(\theta)=\frac{\sqrt{2}}{2}$. Para $sen(\theta)$ obtenemos que $sen(\theta)=\frac{\sqrt{2}}{2}$.

Reemplazamos en $\vec{b}$
$$\vec{b}=\begin{bmatrix}
       -m_1g\mu_{1}\frac{\sqrt{2}}{2}+m_{1}g\frac{\sqrt{2}}{2} \\
       -m_2g\mu_{2}\frac{\sqrt{2}}{2}+m_2g\frac{\sqrt{2}}{2}\\
       -m_3g\mu_{3}\frac{\sqrt{2}}{2}+m_3g\frac{\sqrt{2}}{2}\\
       m_4g
       \end{bmatrix}
$$

Ordenando todo tenemos que resolver el siguiente sistema.
$$A=\begin{bmatrix}
        1& 0 & 0 & -m_1 \\
        -1& 1& 0 &-m_2\\
        0 & -1 & 1 &-m_3 \\
        0 & 0 & 1 & m_4
        \end{bmatrix} \times \begin{bmatrix}
            T_1 \\
            T_2 \\
            T_3 \\
            a \end{bmatrix} = \begin{bmatrix}
       -m_1g\mu_{1}\frac{\sqrt{2}}{2}+m_{1}g\frac{\sqrt{2}}{2} \\
       -m_2g\mu_{2}\frac{\sqrt{2}}{2}+m_2g\frac{\sqrt{2}}{2}\\
       -m_3g\mu_{3}\frac{\sqrt{2}}{2}+m_3g\frac{\sqrt{2}}{2}\\
       +m_4g
       \end{bmatrix}$$
       
Usando el método de Gauss obtenemos que:

Primero definiendo la matriz $A$ y luego el vector $\vec{m}$.

In [65]:
A=np.array([[1,0,0,-10],[-1,1,0,-4],[0,-1,1,-5],[0,0,1,6]])
print(A)

[[  1   0   0 -10]
 [ -1   1   0  -4]
 [  0  -1   1  -5]
 [  0   0   1   6]]


In [66]:
raiz=np.sin(np.pi/4)
g=9.8
mu_1=0.25
mu_2=0.3
mu_3=0.2
masa=np.array([10,4,5,6])
m=np.array([masa[0]*g*raiz*(1-mu_1),masa[1]*g*raiz*(1-mu_2),masa[2]*g*raiz*(1-mu_3),masa[3]*g])
print(m)

[51.97234842 19.40301008 27.71858582 58.8       ]


In [68]:
x = np.linalg.solve(A, m)
print(x)

[35.85477069 48.81074968 68.47054664 -1.61175777]


Entonces obtenemos que la Tension $T_1=35.8547$, $T_2=48.8107$, $T_3=68.4705$ y la aceleración es $a=1.6118[\frac{m}{s^2}]$ hacia abajo.

### Pregunta 2

Tenemos la siguiente matriz A:
    $$A=\begin{bmatrix}
    5+i & 5+2i & -5+3i & 6-3i\\
    5+2i & 7-2i & 8-i & -1+3i \\
    -5+3i & 8-i & -3-3i & 2+2i \\
    6-3i & -1+3i & 2+2i & 8+14i
    \end{bmatrix}$$
    
Y tenemos que resolver el sistema $A\vec{x}=\vec{b}$ con b:
$$\vec{b}=\begin{bmatrix} 15-35i \\ 2+10i \\ -2-34i \\ 8+14i\end{bmatrix}$$

Definimos $A$ y $\vec{b}$:

In [66]:
A=np.array([[5+1j,5+2j,-5+3j,6-3j],[5+2j,7-2j,8-1j,-1+3j],[-5+3j,8-1j,-3-3j,2+2j],[6-3j,-1+3j,2+2j,8+14j]],dtype=np.complex128)
b=np.array([15-35j,2+10j,-2-34j,8+14j],dtype=np.complex128)

Vamos a usar el metodo de Doolitle para resolver la matriz. Primero obtenemos L, y U.

In [152]:
lu, piv = linalg.lu_factor(A)

Luego obtenemos el valor $x$:

In [153]:
x = linalg.lu_solve((lu, piv), b)
print (x)

[ 2.40149288+0.08519672j  0.17494035-4.00400685j -0.57610661+4.08977596j
  0.20984978+0.77490434j]


Podemos ver los valores son diversos, aunque su parte real no son valores grandes todos con menor de 3, su parte imaginario llega a lo más al valor absoluto $4i$. Ahora veamos la precision de la solución.

In [149]:
print (np.dot(A,x)-b)

[ 1.77635684e-15+1.42108547e-14j  3.55271368e-15-3.55271368e-15j
  4.44089210e-16+0.00000000e+00j -1.77635684e-15+3.55271368e-15j]


Vemos que los valores son del orden $10^{-16}$ o $10^{-15}$ que es la precision de la maquina por lo que podemos ver que los valores están acertados.

### Pregunta 3

Suponemos un problema de $N$ incognitas $y_i$. Tambien suponemos que los elementos $A_{ij}$ son de la siguiente forma:
$${a_{ij}=}\frac{1}{i+j-1}$$.

Y el vecto $\vec{b}$ es de la forma:
        $$b_{i}=\frac{1}{i}$$

Utilice alguno de los métodos para obtener la solucion de esta matriz para $N=10$,$N=100$ y $N=1000$.

Tenemos que la solucion exacta $y_i$ es:
            $$\vec{y}=\begin{pmatrix}
            1 \\
            0 \\
            0 \\
            ... \\
            0
            \end{pmatrix}$$

#### Respuesta:

Si analizamos un poco la matriz podemos darnos cuenta que los valores de la matriz son simetricos, es decir que si tenemos $a_{ij}$ es lo mismo que $a_{ji}$. Esto se puede notar porque el valor de $a_{ij}$
$$a_{ij}=\frac{1}{i+j-1}$$

Dado que los reales son conmutativos da lo mismo sumar $i+j$ que sumar $j+i$. Por lo que si lo reordenamos obtenemos el valor de $a_{ij}$.
$$a_{ij}=\frac{1}{i+j-1}=\frac{1}{j+i-1}=a_{ji}$$

Buscamos el valor tal que esta ecuación $i+j-1$ no sea cero, puesto que en python el indice parte
de 0 no de 1. Así que buscamos los valores y desplazamos el indice para evitar que esto ocurra.

Entonces tenemos la siguiente desigualdad que debemos cumplir:
$$i+j-1>0$$
Esto se vuelve cero cuando cuando $i=1$ y $j=0$ o bien $i=0$ y $j=1$. En resumen tenemos que evitar que $i+j=1$ 

Para hacer esto el valor $l$ debe ser mayor que 1, pero dado que le restamos este uno quedando en cero, entonces tenemos que sumarle dos para que siempre sea 1 o mas, nuestra ecuación quedando:
$$i+j-1+2=i+j+1$$

Ahora definimos dos funciones, una función que genera la matriz $A$ y otra que genere el vector $\vec{b}$.

In [79]:
def matrizA(N):
    A=np.zeros((N,N))
    for i in range(N):
        for j in range(N):
            A[i,j]=1/(i+j+1)
    return A

In [80]:
A=matrizA(10)
print(A)

[[ 1.0000  0.5000  0.3333  0.2500  0.2000  0.1667  0.1429  0.1250  0.1111
   0.1000]
 [ 0.5000  0.3333  0.2500  0.2000  0.1667  0.1429  0.1250  0.1111  0.1000
   0.0909]
 [ 0.3333  0.2500  0.2000  0.1667  0.1429  0.1250  0.1111  0.1000  0.0909
   0.0833]
 [ 0.2500  0.2000  0.1667  0.1429  0.1250  0.1111  0.1000  0.0909  0.0833
   0.0769]
 [ 0.2000  0.1667  0.1429  0.1250  0.1111  0.1000  0.0909  0.0833  0.0769
   0.0714]
 [ 0.1667  0.1429  0.1250  0.1111  0.1000  0.0909  0.0833  0.0769  0.0714
   0.0667]
 [ 0.1429  0.1250  0.1111  0.1000  0.0909  0.0833  0.0769  0.0714  0.0667
   0.0625]
 [ 0.1250  0.1111  0.1000  0.0909  0.0833  0.0769  0.0714  0.0667  0.0625
   0.0588]
 [ 0.1111  0.1000  0.0909  0.0833  0.0769  0.0714  0.0667  0.0625  0.0588
   0.0556]
 [ 0.1000  0.0909  0.0833  0.0769  0.0714  0.0667  0.0625  0.0588  0.0556
   0.0526]]


In [66]:
def vectorB(N):
    b=np.zeros(N,dtype=np.double)
    for i in range(N):
        b[i]=1/(i+1)
    return b

In [88]:
b=vectorB(10)
print(b)

[ 1.0000  0.5000  0.3333  0.2500  0.2000  0.1667  0.1429  0.1250  0.1111
  0.1000]


Entonces resolvemos el sistema usando el metodo de Doolittle.

#### Para $N=10$

In [68]:
A=matrizA(10)
b=vectorB(10)

In [93]:
lu, piv = linalg.lu_factor(A)
z=linalg.lu_solve((lu,piv),b)
print(z)

[ 1.0000  0.0000  0.0000  0.0000 -0.0000  0.0000 -0.0000  0.0000 -0.0000
 -0.0000]


Comparamos ahora el valor obtenido con el valor dado por el enunciado.

In [85]:
y=np.zeros(10)
y[0]=1
print(y)

[ 1.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000]


Comparamos la respuesta con el Y dado por el enunciado y vemos que:

In [71]:
print(y-z)

[ 0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000]


No hay diferencia alguna que sea importante.

#### Para $N=100$

In [27]:
A_2=matrizA(100)
b_2=vectorB(100)

In [31]:
lu, piv = linalg.lu_factor(A_2)
z_2=linalg.lu_solve((lu,piv),b_2)
print(z_2)

[ 1.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000 -0.0000
 -0.0000 -0.0000  0.0000  0.0000  0.0000 -0.0000  0.0000 -0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000 -0.0000  0.0000 -0.0000
  0.0000 -0.0000 -0.0000  0.0000  0.0000 -0.0000  0.0000 -0.0000  0.0000
 -0.0000 -0.0000 -0.0000  0.0000 -0.0000 -0.0000  0.0000  0.0000 -0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000 -0.0000 -0.0000  0.0000 -0.0000
  0.0000  0.0000 -0.0000  0.0000  0.0000 -0.0000  0.0000 -0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000 -0.0000 -0.0000 -0.0000 -0.0000
  0.0000 -0.0000  0.0000  0.0000  0.0000 -0.0000 -0.0000  0.0000 -0.0000
  0.0000  0.0000 -0.0000 -0.0000 -0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000 -0.0000 -0.0000  0.0000 -0.0000 -0.0000 -0.0000
 -0.0000]


Nuevamente comparamos el valor obtenido con el valor dado.

In [74]:
y=np.zeros(100,dtype=np.double)
y[0]=1
print(y-z_2)

[ 0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000]


Obtenemos lo mismo que en el caso de $N=10$, no hay alguna diferencia importante.

#### Para $N=1000$

In [29]:
A_3=matrizA(1000)
b_3=vectorB(1000)

In [36]:
lu, piv = linalg.lu_factor(A_3)
z_3=linalg.lu_solve((lu,piv),b_3)
print(z_3)

[ 1.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000 -0.0000
  0.0000 -0.0000  0.0000  0.0000  0.0000 -0.0000 -0.0000 -0.0000  0.0000
 -0.0000 -0.0000 -0.0000  0.0000 -0.0000 -0.0000 -0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000 -0.0000  0.0000 -0.0000  0.0000
  0.0000  0.0000 -0.0000  0.0000  0.0000  0.0000  0.0000  0.0000 -0.0000
  0.0000 -0.0000 -0.0000  0.0000 -0.0000 -0.0000  0.0000  0.0000 -0.0000
  0.0000 -0.0000 -0.0000 -0.0000  0.0000  0.0000  0.0000 -0.0000  0.0000
  0.0000  0.0000 -0.0000 -0.0000  0.0000  0.0000  0.0000  0.0000 -0.0000
  0.0000  0.0000  0.0000 -0.0000  0.0000 -0.0000  0.0000 -0.0000 -0.0000
  0.0000 -0.0000 -0.0000 -0.0000 -0.0000  0.0000 -0.0000 -0.0000 -0.0000
 -0.0000  0.0000  0.0000 -0.0000 -0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000 -0.0000  0.0000 -0.0000 -0.0000 -0.0000  0.0000
  0.0000  0.0000 -0.0000  0.0000  0.0000  0.0000 -0.0000 -0.0000  0.0000
  0.0000 -0.0000 -0.0000 -0.0000 -0.0000  0.0000 -0

Comparamos con el valor dado y obtenemos que:

In [81]:
y=np.zeros(1000)
y[0]=1
print(y-z_3)

[ 0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0

En ninguno de los casos hay diferencias entre la respuesta dada y la respuesta obtenida por el metodo de Doolittle.